sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs

```diff
@@ -0,0 +1,352 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Storage.Shared
+{
+    /// <summary>
+    /// Functions like a readable <see cref="MemoryStream"/> but uses an ArrayPool to supply the backing memory.
+    /// This stream support buffering long sizes.
+    /// </summary>
+    internal class PooledMemoryStream : Stream
```

Comment:
This PR does not address issues of holding everything in memory, or any other optimizations. This simply allows users with good enough hardware to not be limited by our SDK implementation.

Ultimately, we shouldn't stop people who have an unseekable stream of the size where this becomes necessary, want parallel network connections for their upload, and have the hardware to support such a task. This stream as a replacement for single arrays is designed to allow for that, and is what my team considers bare minimum for supporting the next service release.

A future PR for optimizations on uploads will include:
- pure streaming when parallelism is set to 1
- optimizations for when files are used instead of raw streams

Reply:
> This simply allows users with good enough hardware to not be limited by our SDK implementation.

I don't understand this part of the comment.  Could you elaborate on that?

> Ultimately, we shouldn't stop people who have an unseekable stream of the size where this becomes necessary, want parallel network connections for their upload, and have the hardware to support such a task. This stream as a replacement for single arrays is designed to allow for that, and is what my team considers bare minimum for supporting the next service release.

I'm confused by this part as well, but maybe answering the first part will make this clear too.

> pure streaming when parallelism is set to 1

What does pure streaming mean?

Reply:
Our old limitation used a single array to store a whole block. Once you break maxint in size, this no longer works (yes technically the api allows for longs but net461 doesn't and also the arraypool api doesn't). MemoryStream also doesn't satisfy our needs because it doesn't hook up to an array pool for optimizations and doesn't actually support long amounts of data either. Interestingly, I found a microsoft published arraypool backed stream, but it also doesn't support long values. So this stream, as it uses multiple arrays, allows us to store these massive blocks in memory should it be necessary.

The argument in the second part is that some people have the hardware to support these massive memory blocks. Our code, using one array per block, didn't allow them to use jumbo blob's full capabilities even though they had the hardware to run it.

By pure streaming I mean not buffering an entire block in memory before uploading it, as the whole point of buffering the entire block is to start uploading it while we begin reading the next block. If we're uploading blocks sequentially, it makes no sense to buffer the entire block beforehand.